iris-dev#19 Viewing Multiple Segments in the Metadata Browser

eholum opened the pull request:
This PR closes

https://github.com/ChandraCXC/iris-dev/issues/19
https://github.com/ChandraCXC/iris-dev/issues/20

And additional deals with making the metadata browser/visualizer component more stable in general.

We include a new class, StackedStarTable, which is essentially a http://www.star.bristol.ac.uk/~mbt/stil/javadocs/uk/ac/starlink/table/ConcatStarTable.html that also maps all the columns in the provided list of startables into a single metadata star table. Users provide a column matcher class which determines whether or not two columns should "line up". In our case we have implemented a single class, the UtypeColumnInfoMatcher, which may or may not need to be improved upon in the future. For now it appears to line data up well in the test cases that I have provided.

This is achieved through wrapping StarTables into several different other star tables provided by both stil and iris. In particular, a ColumnMappingStarTable uses a ColumnPermutedStarTable and ConstantColumns/JoinStarTables to map an existing StarTable's columns into a provided metadata star table.

e.g.

Provided a metadata table with columns, A - B - C - D, and a data table with columns D - B, the ColumnMappingStarTable maps the baseTable columns to the metadata columns, and turns the dataTable to something looking like, * - B - * - D, where the * are empty columns. From there, it is a simple matter to get a PrimaryMetadataTable by simply creating a set by way of a ColumnMetadataStarTable - which takes a list of startables and produces the smallest columninfo set containing all the tables' columns. Important to note that all of this is generic to StarTables, rather than Iris.

The metadatabrowser now uses StackedStarTables in place of regular StarTables in it's data panels to support viewing of multiple segment's data simultaneously.

I also added some work in refactoring IrisStarTables/SegmentStarTables to make them easier to read and more robust to errors. SegmentStarTables are now the primary data holders for all information related to plotting - and can be extended to be more dynamic than they are now when we are actually interested in working with double[] of data. This was done in an effort to make filtering tasks easier, but there is still more work to be done there.

Lastly was some minor clean up to error handling and the addition of a few more unit tests to verify these changes.

jbudynk commented on iris-common/src/main/java/cfa/vo/iris/sed/stil/SegmentStarTable.java:
```diff
@@ -71,191 +82,276 @@ public SegmentStarTable(Segment segment, String id)
+        
+        Iterator<SegmentDataColumn> it = columns.iterator();
+        for (int i=0; i<index-1; i++) it.next();
+        return it.next();
```
what happens if the index is outside the range of columns?

eholum replied:
Ah, excellent points. i'll add a check and exception of the index is greater than the number of columns. There's no check in a column star table either (https://github.com/Starlink/starjava/blob/83a1eef2bb85ff87772b688803cec44f98745665/table/src/main/uk/ac/starlink/table/ColumnStarTable.java#L94), so I'll change the code that uses those to have a check as well.

jbudynk commented on iris-common/src/main/java/cfa/vo/iris/sed/stil/SegmentStarTable.java:
```diff
@@ -71,191 +82,276 @@ public SegmentStarTable(Segment segment, String id)
         if (specValues == null) return;
         
+        // Convert units
         if (fluxValues != null) {
```
why are the flux-related variables checked for 'not null', but not the spectral-related variables?

eholum replied:
I think because of a prior implementation of units managment that would have thrown a null check if the y values were null. That doesn't seem to be the case anymore - I removed the checks.

jbudynk commented on iris-common/src/main/java/cfa/vo/iris/sed/stil/SegmentStarTable.java:
```diff
@@ -71,191 +82,276 @@ public SegmentStarTable(Segment segment, String id)
 
-    void setFluxValues(double[] fluxValues) {
-        removeColumn(Column.FLUX_COL);
+    public double[] getFluxValues() {
```
the rest of the getters/setters are public, unlike the spectral-related values (and the comment near line 234). is this on purpose?

eholum replied:
Time conflict between this PR and the next one. They should all be public, and I adjusted the comment.

jbudynk commented on iris-visualizer/src/main/java/cfa/vo/iris/visualizer/metadata/MetadataBrowserMainView.java:
```diff
@@ -108,61 +118,109 @@ private void setDataTables() {
+        
+        for (IrisStarTable table : selectedStarTables) {
+            plotterDataTables.add(table.getPlotterTable());
+            segmentDataTables.add(table.getDataTable());
```
I wonder if it makes sense to rename the IrisStarTable get/setDataTable() methods to get/setSegmentDataTable(), given the naming conventions used in this file? In other words, when I see `table.getDataTable()`, I don't think "Segment metadata," I think "spectral, flux, and error values"

eholum replied:
Changed

jbudynk commented on iris-visualizer/src/test/java/cfa/vo/iris/visualizer/stil/tables/VisualizerStarTableTest.java:
```diff
@@ -0,0 +1,41 @@
+import cfa.vo.sedlib.common.Utypes;
+import uk.ac.starlink.table.ColumnInfo;
+
+public abstract class VisualizerStarTableTest {
```
there are no tests run here. It'd be interesting to see a testcase for UTypes that have a different prefix (e.g., handling the case of `spec:SpectralAxis.foo.bar` vs `phot:SpectralAxis.foo.bar`).

eholum replied:
This is just an abstract test case for sharing/setting up common data between some of the other tests in the package.

I'll add a test for it, we have a task in our backlog for finding additional test files that cover the ranges of issues we're likely to see. Also SedLib doesn't include equality checks for prefixes that aren't "spec", unfortunately, I had something in place before to override that - and I can put that back in so that those cases line up.

eholum replied:
Also there wasn't a task for the invalid files list, but I added it. https://github.com/ChandraCXC/iris-dev/issues/50

jbudynk commented:
I've finished reviewing the PR. Looks good for the most part! Definitely more clever than Iris' current metadata browser :) 

Most of my comments you've already addressed (thanks)

I did some testing just fooling around with datasets in the MB. [These are the datasets I used](https://drive.google.com/open?id=0BwbX0VxOVx-iRG1GVmVmTnpGa1U) (Google Drive link).

**RESULTS**
The performance is good. I don't notice any lag in the MB.


**BUGS**
One bug I noticed:  If I make the MB smaller in width, I can't scroll left-and-right in the Data tabs. The same thing happens in the Point Metadata tab, which is weird since I see a scroll bar, but I can't scroll all the way to the right side, and the vertical scroll bar disappears.

**CONCERNS**
I have a few concerns, though I'm not sure they belong to this PR or a previous one. 

1. The column names of the Point Metadata tab change to whatever the actual column names are of the *top-most selected segment*. I would think the x, y, and yerr column names should be either the IVOA standard column names (e.g., the UTypes), or like the names that are in the Data tab (Spectral_Value, Flux_Value, Flux_Error_Value). For example, if you load in the test data set, switch to the Point Metadata tab, and select the J113007.5-144927 segments, you'll find the column names are not informative.

2. The stacked StarTables aren't sorted by spectral value. I don't think that was a requirement we had at the creation of issue iris-dev#19, so I don't know if this needs to be done in this PR. However, it would make great sense if the rows were sorted by spectral value. It could be added in for or after PR #269.

3. Again, perhaps not for this task, but sorting the rows by clicking on a column would be a great and useful addition. (for this, PR #269, or after)

jbudynk commented:
Hmm... I also noticed that something seems off with the Segment tab. With the dataset used above, I see that a number of columns aren't populated at all. Examples are:

[from datasets created in, exported from, and reloaded back into Iris]
 - Extent
 - Min
 - Max
 - StatError

[from NED data; also IVOA UTypes]
 - SpatialAxisCoverageBoundsExtent
 - SpatialAxisCoverageBoundsExtent
 - TimeAxisCoverageBoundsExtent
 - SpectralAxisCoverageBoundsExtent
 - SpectralAxisCoverageBoundsStart / Stop

[from ASDC data]
 - SpaceFrameEquinox

Shouldn't these be cut-off from the data?

I'd also like to note that I tested the MB updates when I remove segments from the SedBuilder, and when I switch between SEDs.

eholum commented:
Re: the resizing bug -
I have added a change to address that in https://github.com/ChandraCXC/iris/pull/269, there were additional changes there that I didn't want to have to deal with in the new layout twice.


Re: 1 -
I changed the behavior of both the point metadata tab and the segment metadata tab to use utypes as column names, when available. This makes those datasets (and most datasets) more clear, and gives us an easy way to remove the behavior if we don't like it at a later time.

Re: 2 and 3 - 
I'm adding tasks for them, as they are beyond the scope of this change:
https://github.com/ChandraCXC/iris-dev/issues/51
https://github.com/ChandraCXC/iris-dev/issues/52

Re: The empty columns in the segment metadata tab-
I added a small bit of logic to remove empty parameters from the segment serialization code. I don't know why sedlib/stil are adding them, but they were easy enough to take out. There are no long null columns in that data tab.

jbudynk commented:
I like the changes. We might want to revisit the names for the x, y, and yerr columns in the Point Metadata tab at some point (they are a little long), but it's mounds better than it was before, especially for those with column names ID_1, ID_2, etc.

I say merge it
